### PR TITLE
Add option to use renderItems with a wrapper for each list element

### PR DESCRIPTION
**Pull request by misspia**
Can update docs in `muban` once merged

Allow `renderItems` to accept a 4th optional argument in the form of an `HTML Element` that will wrap around each handlebar template rendered

Sample usage:

```js
const template = '<div class="bar">{{ text }}</div>';
const wrapper = document.createElement('div');
wrapper.appendAttribute('class', 'foo');

renderItems(container, template, [{text: 'media'}, {text: 'monks'}], false, wrapper);
```

Expected output
```html
<div class="container">
  <div class="foo">
    <div class="bar">
      media
    </div>
  </div>
  <div class="foo">
    <div class="bar">
      monks
    </div>
  </div>
</div>
```

### CODE REVIEW

File: test/dataUtils.spec.ts [hunk 1]
```diff
@@ -245,5 +245,77 @@ describe('dataUtils', () => {
...
+
+      const foo = class Foo {
+        static displayName: string = 'foo';
+
+        constructor() {
+          // dom ready
+          mountSpy('foo');
+        }
+
+        adopted() {
```
Comment (ThaNarie) [inline test/dataUtils.spec.ts]: I think we can remove the adopted and dispose spies for this specific test. The mount spy could still be useful to check if the component is successfully created.

File: test/dataUtils.spec.ts [hunk 2]
```diff
@@ -245,5 +245,77 @@ describe('dataUtils', () => {
...
+        </div>
+      `);
+
+      initComponents(div);
+
+      const wrapperElement = createHTML('<div data-wrapper="bar"></div>');
+      
+      const data = [
+        { text: 'foobar' },
+        { text: 'baz' },
```
Comment (ThaNarie) [inline test/dataUtils.spec.ts]: I would add a 3rd item here, so we know the count is different between the initial HTML and the replaced HTML.

File: src/lib/utils/dataUtils.ts [hunk 3]
```diff
@@ -15,8 +15,23 @@ export function renderItems<T extends Element = HTMLElement>(
   template: (data?: any) => string,
   data: Array<any>,
   append: boolean = false,
+  listElementWrapper?: HTMLElement,
 ) {
-  return render<T>(container, append, () => data.reduce((html, d) => html + template(d), ''));
+  let generateStringTemplate;
+  if (listElementWrapper) {
```
Comment (ThaNarie) [inline src/lib/utils/dataUtils.ts]: I like how this implementation keeps it consistent with the existing code, working with strings.
While the strings were fine before when dealing with just hbs templates, now that we introduce actual DOM elements in between, I feel we have too much back and forth between strings and DOM elements;

1. render hbs template to string
2. convert to DOM with `.innerHTML`
3. convert to string with `.outerHTML`
4. convert to DOM with with `.innerHTML` (in `render`)

I think we could remove step 3 and 4, and move the `.innerHTML` code into the `renderItem(s)`, potentially using a small helper function to reduce duplication. That would be mostly this piece of code:
```ts
const fragment = document.createDocumentFragment();
const div = document.createElement('div');
div.innerHTML = getHtml();
//
fragment.appendChild(child);
```

Let me know if you are able to pick it up, like to me to explain some more, or if you want me to do it :)

File: src/lib/utils/dataUtils.ts [hunk 4]
```diff
@@ -19,21 +19,109 @@ export function renderItems<T extends Element = HTMLElement>(
 ) {
   let generateStringTemplate;
   if (listElementWrapper) {
-    generateStringTemplate = () =>
-      data.reduce((html, d) => {
+    generateStringTemplate = () => {
+      const fragment = document.createDocumentFragment();
+      const div = document.createElement('div');
+      div.innerHTML = generateStringTemplate();
```
Comment (ThaNarie) [inline src/lib/utils/dataUtils.ts]: Wouldn't this result in infinite recursion, calling itself over and over again? :)

Comment (misspia) [reply src/lib/utils/dataUtils.ts]: Ahh sorry, I'm not done yet, just pushed what I had at the end of the day

File: test/dataUtils.spec.ts (hunk 1 above)
Comment (misspia) [reply test/dataUtils.spec.ts]: Removing the dispose spy gives the following error when the test is run:
```
TypeError: removedComponent.instance.dispose is not a function
```

Comment (ThaNarie) [reply test/dataUtils.spec.ts]: The dispose function could stay, the spy can go away.

But maybe better to do a check on this line and only call dispose if it's available:
https://github.com/mediamonks/muban-core/blob/4159a16b2156bb006c9577162a40232801166824/src/lib/utils/cleanElement.ts#L21
(unrelated to this PR, but a nice quick fix to make these tests a bit shorter :))

File: src/lib/utils/dataUtils.ts [hunk 5]
```diff
@@ -7,22 +7,46 @@ export function renderItem<T extends Element = HTMLElement>(
...
   append: boolean = false,
+  listElementWrapper?: HTMLElement,
 ) {
-  return render<T>(container, append, () => data.reduce((html, d) => html + template(d), ''));
+  let generateStringTemplate;
+  if (listElementWrapper) {
+    generateStringTemplate = () => {
+      return data.reduce((html, d) => {
+        const templateHtml = createHtml(template(d), <HTMLElement>listElementWrapper.cloneNode(
+          true,
```
Comment (ThaNarie) [inline src/lib/utils/dataUtils.ts]: Looking at this `deep` clone, it got me thinking that we don't even support Nested wrappers.

If we pass `<div><span></span></div>`, we'll append our rendered template to the outer div (that's the reference we pass), not to the inner span (what we might want).
So for this to work we need to clone the outer element deep, but in the createHtml function, we need to search for the deepest child element to set the innerHTML.

Getting more and more complex :D

Comment (misspia) [reply src/lib/utils/dataUtils.ts]: Haha whoops left the `clone` on that line by accident

Getting more complex but also more interesting :)

**Comment by misspia** [comment]
Looks good! Thanks for adding more types and cleaning up the spies :)

**Pull request merged**